### PR TITLE
Support "v" prefixes in prerelease module version constraints

### LINKED ISSUE #2117: `tofu init` cannot find module version with a "v" prefix
### Thank you for reporting an issue.

Trying to use module `Azure/avm-ptn-alz/azurerm` I have a strange behavior. Without a version constraint it will always use version `0.8.1`

Trying specifing the version manually to `v0.9.0-beta2` will result in the following error:

```text
╷
│ Error: Unresolvable module version constraint
│ 
│   on main.tf line 8:
│    8: module "alz_architecture" {
│ 
│ There is no available version of module "registry.opentofu.org/Azure/avm-ptn-alz/azurerm" (main.tf:8) which matches the given version constraint. The newest available version is 0.8.1.
╵
```

Looking into the `azurerm.json` in the module catalog the beta-versions are available (only an excerpt from the file):

```json
{
  "versions": [
    {
      "version": "v0.9.0-beta2"
    },
    {
      "version": "v0.9.0-beta"
    },
    {
      "version": "v0.8.1"
    }]
}
```

### OpenTofu Version

```shell
OpenTofu v1.8.2
on linux_amd64
```

### OpenTofu Configuration Files

```hcl
module "alz_architecture" {
  source  = "Azure/avm-ptn-alz/azurerm"
  version = "v0.9.0-beta"

  architecture_name  = "alz"
  parent_resource_id = data.azapi_client_config.current.tenant_id
  location           = local.location
}
```

### PR DESCRIPTION
**Pull request by AYM1607**
Resolves #2117

## Target Release

1.9.0

## Checklist

- [x] I have read the [contribution guide](https://github.com/opentofu/opentofu/blob/main/CONTRIBUTING.md).
- [x] I have not used an AI coding assistant to create this PR.
- [x] I have written all code in this PR myself OR I have marked all code I have not written myself (including modified code, e.g. copied from other places and then modified) with a comment indicating where it came from.
- [x] I (and other contributors to this PR) have not looked at the Terraform source code while implementing this PR.

### Go checklist

 

- [x] I have run golangci-lint on my change and receive no errors relevant to my code.
- [x] I have run existing tests to ensure my code doesn't break anything.
- [x] I have added tests for all relevant use cases of my code, and those tests are passing.
- [x] I have only exported functions, variables and structs that should be used from other packages.
- [x] I have added meaningful comments to all exported functions, variables, and structs.

### CODE REVIEW

File: internal/initwd/module_install.go
Comment (AYM1607) [inline internal/initwd/module_install.go:527]: @apparentlymart some tests are succeeding with false positives because we're not using `MeetingConstraintsStringRuby`, see the test log output below:

```
=== RUN   TestModuleInstaller_Prerelease
=== RUN   TestModuleInstaller_Prerelease/exact_match
    module_install_test.go:299: tempChdir switched to /tmp/terraform-configload2467225038 after copying from testdata/prerelease-constraint-match
=== RUN   TestModuleInstaller_Prerelease/exact_match_v_prefix
    module_install_test.go:299: tempChdir switched to /tmp/terraform-configload3963105967 after copying from testdata/prerelease-constraint-v-match
2024-11-05T21:43:08.353Z [WARN]  ModuleInstaller: acctest (while evaluating 0.0.3-alpha.1) attempting to strip "v" prefixes from constraint (v0.0.3-alpha.1) because it was unable to be parsed: a "v" prefix should not be used when specifying versions
=== RUN   TestModuleInstaller_Prerelease/exact_match_eq_selector
    module_install_test.go:299: tempChdir switched to /tmp/terraform-configload4142292899 after copying from testdata/prerelease-constraint-eq-match
=== RUN   TestModuleInstaller_Prerelease/exact_match_v_prefix_eq_selector
    module_install_test.go:299: tempChdir switched to /tmp/terraform-configload3658284982 after copying from testdata/prerelease-constraint-v-eq-match
2024-11-05T21:43:10.049Z [WARN]  ModuleInstaller: acctest (while evaluating 0.0.3-alpha.1) attempting to strip "v" prefixes from constraint (=v0.0.3-alpha.1) because it was unable to be parsed: no spaces allowed after operator "="
=== RUN   TestModuleInstaller_Prerelease/partial_match
    module_install_test.go:299: tempChdir switched to /tmp/terraform-configload581610588 after copying from testdata/prerelease-constraint
=== RUN   TestModuleInstaller_Prerelease/partial_match_v_prefix
    module_install_test.go:299: tempChdir switched to /tmp/terraform-configload3386838246 after copying from testdata/prerelease-constraint-v
2024-11-05T21:43:11.952Z [WARN]  ModuleInstaller: acctest (while evaluating 0.0.3-alpha.1) attempting to strip "v" prefixes from constraint (<=v0.0.3-alpha.1) because it was unable to be parsed: no spaces allowed after operator "<="
=== RUN   TestModuleInstaller_Prerelease/multiple_constraints
    module_install_test.go:299: tempChdir switched to /tmp/terraform-configload1008194909 after copying from testdata/prerelease-constraint-multiple
2024-11-05T21:43:12.838Z [WARN]  ModuleInstaller: acctest (while evaluating 0.0.3-alpha.1) attempting to strip "v" prefixes from constraint (0.0.3-alpha.1, >=0.0.2) because it was unable to be parsed: commas are not needed to separate version selections; separate with spaces instead
2024-11-05T21:43:12.838Z [WARN]  ModuleInstaller: acctest ignoring 0.0.3-alpha.1 because the version constraints (0.0.3-alpha.1, >=0.0.2) could not be parsed: commas are not needed to separate version selections; separate with spaces instead
=== RUN   TestModuleInstaller_Prerelease/err
    module_install_test.go:299: tempChdir switched to /tmp/terraform-configload1249799817 after copying from testdata/prerelease-constraint-err
=== RUN   TestModuleInstaller_Prerelease/err_v_prefix
    module_install_test.go:299: tempChdir switched to /tmp/terraform-configload2000343409 after copying from testdata/prerelease-constraint-v-err
2024-11-05T21:43:13.047Z [WARN]  ModuleInstaller: acctest (while evaluating 0.0.3-alpha.1) attempting to strip "v" prefixes from constraint (>=v0.0.3-alpha.1) because it was unable to be parsed: no spaces allowed after operator ">="
--- PASS: TestModuleInstaller_Prerelease (6.23s)
    --- PASS: TestModuleInstaller_Prerelease/exact_match (1.06s)
    --- PASS: TestModuleInstaller_Prerelease/exact_match_v_prefix (1.23s)
    --- PASS: TestModuleInstaller_Prerelease/exact_match_eq_selector (0.65s)
    --- PASS: TestModuleInstaller_Prerelease/exact_match_v_prefix_eq_selector (1.04s)
    --- PASS: TestModuleInstaller_Prerelease/partial_match (0.84s)
    --- PASS: TestModuleInstaller_Prerelease/partial_match_v_prefix (1.08s)
    --- PASS: TestModuleInstaller_Prerelease/multiple_constraints (0.11s)
    --- PASS: TestModuleInstaller_Prerelease/err (0.10s)
    --- PASS: TestModuleInstaller_Prerelease/err_v_prefix (0.11s)
```

I'd expect to see a log indicating that the constraints are not met for the "multiple constraints" test, but instead I see "commas are not needed to separate version selections; separate with spaces instead".

What do you think about creating another issue to investigate changing to the ruby-style parser?

Comment (apparentlymart) [reply internal/initwd/module_install.go:542]: Thanks for pointing this out, @AYM1607.

While this is unfortunate, I think we should leave this "broken" for now because fixing it would potentially make a future version of OpenTofu select a prerelease version where an earlier version would not have, and that could have a harmful outcome.

Once we've got the immediate bug fixed (by merging this PR) I'll make a new issue to record the compromise we made here, as you suggested. I also pushed a final commit to your branch to add some more commentary in the code itself to try to help a future maintainer who will inevitably wonder why the logic is structured in this way.

(I also took the opportunity to fix the merge conflict on the changelog while I was adding a new commit anyway, to save us from having to do another review round-trip, since I caused that conflict with some work yesterday to prepare the changelog for the forthcoming v1.9.0 release.)

I'm intending to merge this once the checks are all passing.

**Review by apparentlymart — APPROVED** [review]
Thanks for working on this @AYM1607, and in particular for your patience while we worked through different possible approaches to fix it.

Although what we've ended up here is definitely not what we would've written if we were writing this code new today, I think it represents the best compromise to narrowly fix just the bug that was reported in https://github.com/opentofu/opentofu/issues/2117 while minimizing the risk of changing OpenTofu's interpretation of any other version constraint string.

Once this is merged (which I intend to do once the checks all show as passing) I will make a new issue to represent both the existing technical debt and the new technical debt we've accepted in this PR, so that we'll have a record of it and consider how we might make a more comprehensive improvement to this component in future.

Thanks again!

**Review by apparentlymart — APPROVED** [review]

**Pull request merged**
